core/blockchain.go

```diff
@@ -3444,36 +3465,86 @@ func (bc *BlockChain) PreviousCanonicalCoincidentOnPath(header *types.Header, sl
 // dominant block is canonical.
 func (bc *BlockChain) CheckCanonical(header *types.Header, order int) error {
 	lastUncleHeader := &types.Header{}
+	lastUncleHash := common.Hash{}
 	for {
 		status := bc.domClient.GetBlockStatus(context.Background(), header)
 		// If the header is cononical break else keep looking
 		switch status {
 		case quaiclient.CanonStatTy:
-			if (lastUncleHeader != &types.Header{}) {
+			if (lastUncleHash != common.Hash{}) {
+				fmt.Println("header", header.Number, header.Hash(), lastUncleHeader.Number, lastUncleHash)
 				bc.ReOrgRollBack(lastUncleHeader, []*types.Header{}, []*types.Header{})
+				return consensus.ErrSubordinateNotSynced
 			}
 			return nil
 		default:
 			lastUncleHeader = header
+			lastUncleHash = lastUncleHeader.Hash()
 		}
 
 		if header.Number[types.QuaiNetworkContext].Cmp(big.NewInt(0)) == 0 {
+			if (lastUncleHeader != &types.Header{}) {
+				return consensus.ErrSubordinateNotSynced
+			}
 			return nil
 		}
 
 		terminalHeader, err := bc.Engine().PreviousCoincidentOnPath(bc, header, header.Location, order, types.QuaiNetworkContext, true)
 		if err != nil {
+			log.Warn("Error in PCOP", "err", err)
 			return err
 		}
 
 		if terminalHeader.Number[types.QuaiNetworkContext].Cmp(big.NewInt(0)) == 0 {
+			if (lastUncleHeader != &types.Header{}) {
+				return consensus.ErrSubordinateNotSynced
+			}
 			return nil
 		}
 
 		header = terminalHeader
 	}
 }
 
+// CheckSubordinateHeader asks the subordinate chain what their header is and will manifest
+// the subordinate chain to the proper head.
+func (bc *BlockChain) CheckSubordinateHeader(header *types.Header) error {
+	subClient := bc.subClients[header.Location[types.QuaiNetworkContext]-1]
+	// Do not validate PCRC on a subclient you do not have.
+	if subClient == nil {
+		return nil
+	}
+
+	subHead, err := subClient.HeaderByNumber(context.Background(), nil)
+	if err != nil {
+		return err
+	}
+	// If sub is not ready to run PCRC on a dominant block, sync the subordinate chain.
+	if subHead.Hash() != header.ParentHash[types.QuaiNetworkContext+1] {
+		extBlocks, err := bc.GetExternalBlockTraceSet(subHead.Hash(), header, types.QuaiNetworkContext+1)
```

Comment:
If I read this sync logic correctly, this only works if the sub is behind the header. What happens if the sub is on an uncle chain? `subHead.Hash()` would not match `header.ParentHash[...]`, but no amount of syncing would get the sub on the right head. The sub would need to reorg and then sync. Or maybe I'm thinking about this wrong?